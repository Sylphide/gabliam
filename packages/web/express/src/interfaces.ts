import { express } from './express';
import { defaultMethods } from '@gabliam/web-core';

/**
 * Represent a method that create an express router
 */
export type RouterCreator = () => express.Router;

export type ExpressMethods = Extract<keyof express.Router, defaultMethods>;
