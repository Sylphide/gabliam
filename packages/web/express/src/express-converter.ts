import {
  Container,
  InjectContainer,
  INJECT_CONTAINER_KEY,
  Service,
  toPromise,
} from '@gabliam/core';
import {
  createInterceptorResolver,
  extractParameters,
  getParameterMetadata,
  isInterceptor,
  getContext,
} from '@gabliam/web-core';
import { express } from './express';

/**
 * Class for convert interceptor to Express Middleware
 */
@InjectContainer()
@Service()
export class ExpressConverter {
  public interceptorToMiddleware(clazz: any): express.RequestHandler {
    const container = <Container>(this as any)[INJECT_CONTAINER_KEY];
    const instance = createInterceptorResolver(container)(clazz);
    if (!isInterceptor(instance)) {
      throw new Error();
    }

    return async (
      req: express.Request,
      res: express.Response,
      next: express.NextFunction
    ) => {
      const args = extractParameters(
        instance,
        'intercept',
        null,
        getContext(req),
        next,
        getParameterMetadata(instance, 'intercept')
      );

      try {
        await toPromise((instance['intercept'] as any)(...args));
        next();
      } catch (err) {
        next(err);
      }
    };
  }
}
