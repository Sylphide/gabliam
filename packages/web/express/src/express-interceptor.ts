import {
  AfterResponseInterceptor,
  Interceptor,
  InterceptorMethod,
  InterceptorConstructor,
  UseInterceptors,
} from '@gabliam/web-core';
import { METADATA_KEY } from './constants';
import { express } from './express';
import { Service } from '@gabliam/core';

/**
 * Test if target is an express interceptor
 * @param target any
 */
export const isExpressInterceptor = (target: any) =>
  Reflect.getOwnMetadata(
    METADATA_KEY.expressInterceptor,
    target.constructor || target
  ) === true;

/**
 * ExpressInterceptor decorator
 * class is an express interceptor.
 * Must return express middleware
 */
const ExpressInterceptor = () => (target: any) => {
  Reflect.defineMetadata(METADATA_KEY.expressInterceptor, true, target);
};

/**
 * Convert a Express middleware to an express interceptor
 */
export const toInterceptor = (
  mid: express.RequestHandler | express.ErrorRequestHandler,
  type: InterceptorMethod = 'intercept'
): InterceptorConstructor => {
  let clazz: InterceptorConstructor;
  if (type === 'intercept') {
    clazz = class implements Interceptor {
      intercept() {
        return mid;
      }
    };
  } else {
    clazz = class implements AfterResponseInterceptor {
      afterResponse() {
        return mid;
      }
    };
  }
  Service()(clazz);
  ExpressInterceptor()(clazz);
  return clazz;
};

/**
 * Alias for evict to use: UseInterceptors(toInterceptor(expressMid))
 */
export const UseExpressInterceptors = (
  ...mids: (express.RequestHandler | express.ErrorRequestHandler)[]
) => UseInterceptors(...mids.map(m => toInterceptor(m)));
