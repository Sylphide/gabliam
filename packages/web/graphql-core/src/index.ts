export * from './interfaces';
export { TYPE } from './constants';
export * from './graphql-core-plugin';
export * from './decorator';
export * from 'graphql-tools';
export * from 'graphql';
