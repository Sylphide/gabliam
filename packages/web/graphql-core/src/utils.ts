import { reflection } from '@gabliam/core';
import { GraphQLResolveInfo } from 'graphql';
import { extractArgsFn } from './interfaces';
import { GqlParamDecorator } from './metadatas';

export const getExtractArgs = (
  controller: any,
  propKey: string
): extractArgsFn => {
  const params = reflection.parameters(<any>controller.constructor, propKey);

  if (params.length === 0) {
    return (source: any, args: any, context: any, info: GraphQLResolveInfo) => [
      source,
      args,
      context,
      info,
    ];
  }

  const parameters = <[string | undefined, GqlParamDecorator][]>params.map(
    meta => {
      let type: string | undefined;
      if (meta.length === 2) {
        type = meta[0].name;
      }
      return [type, meta.slice(-1)[0] as GqlParamDecorator];
    }
  );

  return (source: any, args: any, context: any, info: GraphQLResolveInfo) => {
    return parameters.map(([type, p]) =>
      p.handler(p.args, type, source, args, context, info)
    );
  };
};
