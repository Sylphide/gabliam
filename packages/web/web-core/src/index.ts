export * from './constants';
export * from './cookies';
export * from './decorators';
export * from './errors';
export * from './execution-context';
export * from './gab-context';
export * from './gab-request';
export * from './gab-response';
export * from './interface';
export * from './plugin-config';
export * from './response-entity';
export * from './utils';
export * from './validate';
export * from './web-configuration';
export * from './web-plugin-base';
