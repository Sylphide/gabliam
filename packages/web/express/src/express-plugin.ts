import {
  Container,
  GabliamPlugin,
  Plugin,
  Registry,
  Scan,
  toPromise,
} from '@gabliam/core';
import {
  APP,
  cleanPath,
  compose,
  ExecutionContext,
  getContext,
  InterceptorInfo,
  RestMetadata,
  SERVER,
  WebConfiguration,
  WebPluginBase,
  WebPluginConfig,
  WEB_PLUGIN_CONFIG,
} from '@gabliam/web-core';
import * as d from 'debug';
import * as http from 'http';
import { CUSTOM_ROUTER_CREATOR } from './constants';
import { converterValue, send } from './converter-value';
import { express } from './express';
import { isValidInterceptor } from './express-interceptor';
import { ExpressMethods, RouterCreator } from './interfaces';
import { addContextMiddleware, addMiddlewares } from './middleware';

const debug = d('Gabliam:Plugin:ExpressPlugin');

@Plugin('ExpressPlugin')
@Scan()
export class ExpressPlugin extends WebPluginBase implements GabliamPlugin {
  bindApp(
    container: Container,
    registry: Registry,
    webConfiguration: WebConfiguration
  ): void {
    container.bind(APP).toConstantValue(express());
    webConfiguration.addwebConfig({
      instance: addMiddlewares,
      order: -2,
    });

    webConfiguration.addwebConfig({
      instance: addContextMiddleware,
      order: -1,
    });
  }

  async destroy(container: Container, registry: Registry) {
    await this.stop(container, registry);
  }

  async stop(container: Container, registry: Registry) {
    try {
      // server can be undefined (if start is not called)
      const server = container.get<http.Server>(SERVER);
      return new Promise<void>(resolve => {
        server.close(() => resolve());
      });
    } catch (e) {}
  }

  async start(container: Container, registry: Registry) {
    const restConfig = container.get<WebPluginConfig>(WEB_PLUGIN_CONFIG);
    const app = container.get<express.Application>(APP);
    const port = restConfig.port;
    app.set('port', port);

    const server = http.createServer(<any>app);
    server.listen(port, restConfig.hostname);
    server.on('error', onError);
    server.on('listening', onListening);
    container.bind(SERVER).toConstantValue(server);

    /* istanbul ignore next */
    function onError(error: NodeJS.ErrnoException): void {
      // tslint:disable-next-line:curly
      if (error.syscall !== 'listen') throw error;
      const bind = typeof port === 'string' ? 'Pipe ' + port : 'Port ' + port;
      switch (error.code) {
        case 'EACCES':
          console.error(`${bind} requires elevated privileges`);
          process.exit(1);
          break;
        case 'EADDRINUSE':
          console.error(`${bind} is already in use`);
          process.exit(1);
          break;
        default:
          throw error;
      }
    }

    /* istanbul ignore next */
    function onListening(): void {
      const addr = server.address();
      const bind =
        typeof addr === 'string' ? `pipe ${addr}` : `port ${addr.port}`;
      console.log(`Listening on ${bind}`);
    }
  }

  async buildControllers(
    restMetadata: RestMetadata<ExpressMethods>,
    container: Container
  ) {
    const app = container.get<express.Application>(APP);

    // get the router creator
    let routerCreator: RouterCreator = () => express.Router();
    try {
      routerCreator = container.get<RouterCreator>(CUSTOM_ROUTER_CREATOR);
    } catch (e) {}

    for (const [
      controllerId,
      { methods, controllerPath },
    ] of restMetadata.controllerInfo) {
      const controller = container.get<object>(controllerId);
      const router = routerCreator();
      const routerPath = cleanPath(`${restMetadata.rootPath}${controllerPath}`);

      debug(`New route : "${routerPath}"`);

      for (const methodInfo of methods) {
        const execCtx = new ExecutionContext(controller, methodInfo);

        const interceptors = methodInfo.interceptors.filter(i =>
          isValidInterceptor(i)
        );

        // create handler
        const handler = this.handlerFactory(execCtx, interceptors);

        // register handler in router
        router[methodInfo.method](methodInfo.methodPath, handler);
      }
      app.use(routerPath, router);
    }
  }

  private handlerFactory(
    execCtx: ExecutionContext,
    interceptors: InterceptorInfo[]
  ): express.RequestHandler {
    return async (
      req: express.Request,
      res: express.Response,
      next: express.NextFunction
    ) => {
      const composeInterceptor = compose(
        interceptors,
        converterValue
      );
      const ctx = getContext(req);
      const methodInfo = execCtx.getMethodInfo();
      const controller = execCtx.getClass();

      let expressNextCall = false;
      const expressNext = () => {
        expressNextCall = true;
        next();
      };

      const callNext = async () => {
        const args = methodInfo.extractArgs(ctx, execCtx, expressNext);
        return await toPromise(controller[methodInfo.methodName](...args));
      };

      try {
        await composeInterceptor(ctx, execCtx, callNext);
        if (!expressNextCall) {
          send(ctx, res, methodInfo.json);
        }
      } catch (err) {
        next(err);
      }
    };
  }
}
