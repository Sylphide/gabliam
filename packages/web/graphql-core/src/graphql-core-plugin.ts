import {
  Container,
  GabliamPlugin,
  reflection,
  Registry,
  Scan,
  toPromise,
} from '@gabliam/core';
import { SERVER_STARTER } from '@gabliam/web-core';
import { ApolloServerBase } from 'apollo-server-core';
import * as d from 'debug';
import * as fs from 'fs';
import { GraphQLResolveInfo, GraphQLSchema } from 'graphql';
import { IResolvers, makeExecutableSchema } from 'graphql-tools';
import * as _ from 'lodash';
import { DEBUG_PATH, GRAPHQL_CONFIG, METADATA_KEY, TYPE } from './constants';
import { GraphqlServerStarter } from './graphql-server-starter';
import { GraphqlConfig } from './interfaces';
import { GraphqlController, Resolver, ResolverType } from './metadatas';
import { getExtractArgs } from './utils';

const debug = d(DEBUG_PATH);

@Scan()
export abstract class GraphqlCorePlugin implements GabliamPlugin {
  async build(container: Container, registry: Registry) {
    const graphqlPluginConfig = container.get<GraphqlConfig>(GRAPHQL_CONFIG);
    const graphqlServerStarter = container.get<GraphqlServerStarter>(
      SERVER_STARTER
    );
    const listdefinitions: string[] = [];

    if (graphqlPluginConfig.graphqlFiles) {
      listdefinitions.push(
        ...this.loadGraphqlFiles(...graphqlPluginConfig.graphqlFiles)
      );
    }

    const controllerIds = registry.get(TYPE.Controller);
    const resolverList: IResolvers[] = [];

    for (const { id: controllerId } of controllerIds) {
      const controller = container.get<object>(controllerId);

      const [controllerMetadata] = reflection
        .annotationsOfDecorator<GraphqlController>(
          controller.constructor,
          GraphqlController
        )
        .slice(-1);

      listdefinitions.push(...controllerMetadata.schema);
      listdefinitions.push(
        ...this.loadGraphqlFiles(...controllerMetadata.graphqlFiles)
      );

      const resolverMetadatas = reflection.propMetadataOfDecorator<Resolver>(
        controller.constructor,
        METADATA_KEY.resolver
      );

      if (resolverMetadatas && controllerMetadata) {
        for (const [methodName, metas] of Object.entries(resolverMetadatas)) {
          const [resolverMetadata] = metas.slice(-1);
          if (resolverMetadata.schema) {
            listdefinitions.push(resolverMetadata.schema);
          }
          if (resolverMetadata.graphqlFile) {
            listdefinitions.push(
              ...this.loadGraphqlFiles(resolverMetadata.graphqlFile)
            );
          }

          resolverList.push(
            await this.getResolver(controller, methodName, resolverMetadata)
          );
        }
      }
    }

    const queries = [];
    const mutations = [];
    const others = [];
    const subscriptions = [];

    for (const definitions of listdefinitions) {
      const types = definitions
        .split('}')
        .map(s => s.replace(/^\s+|\s+$/g, '')) // remove space
        .filter(s => s !== '}' && s !== '') // remove usless
        .filter(s => s[0] !== '#') // remove commentaries lines
        .map(s => s + '\n}');

      for (let type of types) {
        if (/\s*type\s*Query\s*{\s*/g.test(type)) {
          if (type.slice(0, 6) !== 'extend') {
            // add extend
            type = `extend ${type}`;
          }
          queries.push(type);
        } else if (/\s*type\s*Mutation\s*{\s*/g.test(type)) {
          if (type.slice(0, 6) !== 'extend') {
            type = `extend ${type}`;
          }
          mutations.push(type);
        } else if (/\s*type\s*Subscription\s*{\s*/g.test(type)) {
          if (type.slice(0, 6) !== 'extend') {
            type = `extend ${type}`;
          }
          subscriptions.push(type);
        } else {
          others.push(type);
        }
      }
    }

    const typeDefs: string[] = [];
    // for first Type Query remove extend
    if (queries.length) {
      queries[0] = queries[0].slice(7);
      typeDefs.push(...queries);
    }

    if (mutations.length) {
      mutations[0] = mutations[0].slice(7);
      typeDefs.push(...mutations);
    }

    if (subscriptions.length) {
      subscriptions[0] = subscriptions[0].slice(7);
      typeDefs.push(...subscriptions);
    }

    typeDefs.push(...others);

    const resolvers = <IResolvers>_.merge({}, ...resolverList);

    if (Object.keys(resolvers).length === 0) {
      return;
    }

    debug('typeDefs', typeDefs);
    const schema = <GraphQLSchema>(<any>makeExecutableSchema({
      typeDefs,
      resolvers,
    }));

    const apolloServer = this.getApolloServer(
      container,
      registry,
      graphqlPluginConfig,
      schema
    );
    graphqlServerStarter.apolloServer = apolloServer;
  }

  abstract getApolloServer(
    container: Container,
    registry: Registry,
    graphqlPluginConfig: GraphqlConfig,
    schema: GraphQLSchema
  ): ApolloServerBase;

  private loadGraphqlFiles(...files: string[]) {
    return files.map(file => fs.readFileSync(file, 'UTF-8'));
  }

  private async getResolver(
    instance: any,
    methodName: string,
    resolverMetadata: Resolver
  ): Promise<IResolvers> {
    let resolver: any;

    switch (resolverMetadata.type) {
      case ResolverType.Query:
      case ResolverType.Mutation:
        const extractParameters = getExtractArgs(instance, methodName);
        resolver = (
          source: any,
          args: any,
          context: any,
          info: GraphQLResolveInfo
        ) =>
          instance[methodName](
            ...extractParameters(source, args, context, info)
          );
        break;
      case ResolverType.Subscription:
      case ResolverType.Map:
      case ResolverType.ResolveType:
        resolver = await toPromise(instance[methodName]());
        break;
    }

    return _.set<IResolvers>({}, resolverMetadata.path, resolver);
  }
}
