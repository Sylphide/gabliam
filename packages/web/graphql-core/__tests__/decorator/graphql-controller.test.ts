// tslint:disable:one-line
// tslint:disable:no-unused-expression
import { reflection } from '@gabliam/core';
import * as path from 'path';
import { GraphqlController } from '../../src';
import { HeroSchema, PhotoSchema } from '../fixtures/schemas/schema';

test('should add GraphqlController metadata to a class when decorated with @GraphqlController()', () => {
  @GraphqlController()
  class TestBean {}

  expect(reflection.annotations(TestBean)).toMatchSnapshot();
});

test('should add GraphqlController metadata to a class when decorated with @GraphqlController({graphqlFiles: absolute})', () => {
  @GraphqlController({
    graphqlFiles: [
      `${__dirname}/photo/hero.gql`,
      `${__dirname}/photo/photo.gql`,
    ],
  })
  class TestBean {}

  expect(reflection.annotations(TestBean)).toMatchSnapshot();
});

test('should add GraphqlController metadata to a class when decorated with @GraphqlController({graphqlFiles: relative})', () => {
  @GraphqlController({
    graphqlFiles: [`./photo/hero.gql`, `./photo/photo.gql`],
  })
  class TestBean {}

  expect(reflection.annotations(TestBean)).toMatchSnapshot();
});

test('should add GraphqlController metadata to a class when decorated with @GraphqlController({graphqlFiles: relative, pwd})', () => {
  @GraphqlController({
    graphqlFiles: [`./photo/hero.gql`, `./photo/photo.gql`],
    pwd: path.resolve(__dirname, '..'),
  })
  class TestBean {}

  expect(reflection.annotations(TestBean)).toMatchSnapshot();
});

test('should add GraphqlController metadata to a class when decorated with @GraphqlController({schema})', () => {
  @GraphqlController({
    schema: [PhotoSchema, HeroSchema],
  })
  class TestBean {}

  expect(reflection.annotations(TestBean)).toMatchSnapshot();
});

test('should add GraphqlController metadata to a class when decorated with @GraphqlController({graphqlFiles, schema})', () => {
  @GraphqlController({
    graphqlFiles: [`${__dirname}/photo/photo.gql`],
    schema: [HeroSchema],
  })
  class TestBean {}

  expect(reflection.annotations(TestBean)).toMatchSnapshot();
});
