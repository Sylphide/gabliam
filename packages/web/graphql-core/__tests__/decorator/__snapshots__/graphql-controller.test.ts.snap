// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`should add GraphqlController metadata to a class when decorated with @GraphqlController() 1`] = `
Array [
  DecoratorFactory {
    "graphqlFiles": Array [],
    "schema": Array [],
  },
  DecoratorFactory {
    "id": [Function],
    "type": "GraphqlController",
  },
]
`;

exports[`should add GraphqlController metadata to a class when decorated with @GraphqlController({graphqlFiles, schema}) 1`] = `
Array [
  DecoratorFactory {
    "graphqlFiles": Array [
      "<PROJECT_ROOT>/packages/web/graphql-core/__tests__/decorator/photo/photo.gql",
    ],
    "schema": Array [
      "
type Hero {
  id: Int!

  name: String!

  power: String!

  amountPeopleSaved: Int!
}

input HeroInput {
  name: String!

  power: String!

  amountPeopleSaved: Int!
}

type HeroPage {
  items: [Hero]
  totalCount: Int
}

type Query {
  heroes: [Hero]
  getPageOfHeroes(page: Int, perPage: Int, sortField: String, sortOrder: String, filter: String): HeroPage
}

type Mutation {
  submitHero(heroInput: HeroInput!): Hero
}
",
    ],
  },
  DecoratorFactory {
    "id": [Function],
    "type": "GraphqlController",
  },
]
`;

exports[`should add GraphqlController metadata to a class when decorated with @GraphqlController({graphqlFiles: absolute}) 1`] = `
Array [
  DecoratorFactory {
    "graphqlFiles": Array [
      "<PROJECT_ROOT>/packages/web/graphql-core/__tests__/decorator/photo/hero.gql",
      "<PROJECT_ROOT>/packages/web/graphql-core/__tests__/decorator/photo/photo.gql",
    ],
    "schema": Array [],
  },
  DecoratorFactory {
    "id": [Function],
    "type": "GraphqlController",
  },
]
`;

exports[`should add GraphqlController metadata to a class when decorated with @GraphqlController({graphqlFiles: relative, pwd}) 1`] = `
Array [
  DecoratorFactory {
    "graphqlFiles": Array [
      "<PROJECT_ROOT>/packages/web/graphql-core/__tests__/photo/hero.gql",
      "<PROJECT_ROOT>/packages/web/graphql-core/__tests__/photo/photo.gql",
    ],
    "schema": Array [],
  },
  DecoratorFactory {
    "id": [Function],
    "type": "GraphqlController",
  },
]
`;

exports[`should add GraphqlController metadata to a class when decorated with @GraphqlController({graphqlFiles: relative}) 1`] = `
Array [
  DecoratorFactory {
    "graphqlFiles": Array [
      "<PROJECT_ROOT>/packages/web/graphql-core/__tests__/decorator/photo/hero.gql",
      "<PROJECT_ROOT>/packages/web/graphql-core/__tests__/decorator/photo/photo.gql",
    ],
    "schema": Array [],
  },
  DecoratorFactory {
    "id": [Function],
    "type": "GraphqlController",
  },
]
`;

exports[`should add GraphqlController metadata to a class when decorated with @GraphqlController({schema}) 1`] = `
Array [
  DecoratorFactory {
    "graphqlFiles": Array [],
    "schema": Array [
      "
type Photo {
  id: Int!

  name: String!

  description: String!

  fileName: String!

  views: Int!

  isPublished: Boolean!
}

input PhotoInput {
  name: String!

  description: String!

  fileName: String!

  views: Int!

  isPublished: Boolean!
}

type PhotoPage {
  items: [Photo]
  totalCount: Int
}

type Query {
  photos: [Photo]
  getPageOfPhotos(page: Int, perPage: Int, sortField: String, sortOrder: String, filter: String): PhotoPage
}

type Mutation {
  submitPhoto(photoInput: PhotoInput!): Photo
}
",
      "
type Hero {
  id: Int!

  name: String!

  power: String!

  amountPeopleSaved: Int!
}

input HeroInput {
  name: String!

  power: String!

  amountPeopleSaved: Int!
}

type HeroPage {
  items: [Hero]
  totalCount: Int
}

type Query {
  heroes: [Hero]
  getPageOfHeroes(page: Int, perPage: Int, sortField: String, sortOrder: String, filter: String): HeroPage
}

type Mutation {
  submitHero(heroInput: HeroInput!): Hero
}
",
    ],
  },
  DecoratorFactory {
    "id": [Function],
    "type": "GraphqlController",
  },
]
`;
